src/metabase/models/table.clj

```diff
@@ -127,6 +131,73 @@
   ([_ pk]
    (mi/can-write? (t2/select-one :model/Table pk))))
 
+;;; ------------------------------------------------ SQL Permissions ------------------------------------------------
+
+(def ^:private TableVisibilityConfig
+  [:map
+   [:user-id pos-int?]
+   [:is-superuser? :boolean]
+   [:view-data-permission-level data-perms/PermissionValue]
+   [:create-queries-permission-level data-perms/PermissionValue]])
+
+(mu/defn- perm-type-to-int-case
+  [perm-type :- data-perms/PermissionType column-or-perm-value :- [:or :keyword [:tuple [:= ::h2x/literal] :string]]]
+  (into [:case]
+        (apply concat
+               (map-indexed (fn [idx perm-value] [[:= column-or-perm-value (h2x/literal perm-value)] [:inline idx]])
+                            (-> data-perms/Permissions perm-type :values)))))
+
+(mu/defn- perm-condition
+  [perm-type :- :keyword required-level :- :keyword equality-comp]
+  [:and
+   equality-comp
+   [:= :dp.perm_type (h2x/literal perm-type)]
+   [:<=
+    (perm-type-to-int-case perm-type :dp.perm_value)
+    (perm-type-to-int-case perm-type (h2x/literal required-level))]])
+
+(mu/defn- user-in-group-half-join
+  [user-id :- pos-int?]
+  [:exists {:select [1]
+            :from   [[:permissions_group :pg]]
+            :where  [:and
+                     [:= :pg.id :dp.group_id]
+                     [:exists {:select [1]
+                               :from [[:permissions_group_membership :pgm]]
+                               :where [:and
+                                       [:= :pgm.group_id :pg.id]
+                                       [:= :pgm.user_id [:inline user-id]]]}]]}])
+
+(mu/defn- has-perms-for-table-as-honey-sql?
+  [user-id :- pos-int? perm-type :- :keyword required-level :- :keyword]
+  [:exists {:select [1]
+            :from   [[:data_permissions :dp]]
+            :where  [:and [:or
+                           (perm-condition perm-type required-level [:and [:= :t.db_id :dp.db_id]
+                                                                     [:= :dp.table_id nil]])
+                           (perm-condition perm-type required-level [:= :t.id :dp.table_id])]
+                     (user-in-group-half-join user-id)]}])
+
+(mu/defn visible-tables-filter-clause
+  "Returns a query for tables that are visible to the supplied user given the supplied view-data & create-queries permission levels."
+  [table-id-field-or-fields :- [:or [:sequential :keyword] :keyword]
```

Comment:
nit - what do we gain from parameterizing this by the table ID field? and why do we need to support multiple field names? I don't see that being used

Reply:
The parameterize is used [here](https://github.com/metabase/metabase/pull/57541/files#diff-02aecb36867031ea129ab28bab7705dc2441dfead5a65b41799f71548e6674abR166). @nvoxland do you know why we have their parameterization for the collection model in legacy search? How is search_index used there?